src/Polysemy/ConstraintAbsorber.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE AllowAmbiguousTypes         #-}
+{-# LANGUAGE ConstraintKinds             #-}
+{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
+
+module Polysemy.ConstraintAbsorber
+  (
+    -- * absorb-builder
+   absorbWithSem
+  
+    -- * Re-exports
+  , Reifies
+  , (:-)(Sub)
+  , Dict(Dict)
+  , reflect
+  , Proxy (Proxy)
+  )
+where
+
+import           Polysemy
+
+import qualified Data.Constraint as C
+import           Data.Constraint (Dict(Dict),(:-)(Sub),(\\))
+import qualified Data.Constraint.Unsafe as C
+import           Data.Proxy (Proxy (..))
+import qualified Data.Reflection as R
+import           Data.Reflection (Reifies, reflect) 
+import           Data.Kind (Type, Constraint)
+
+------------------------------------------------------------------------------
+-- | Given a reifiable constraint @p,
+-- discharge the constraint by providing a dictionary, and a "reified instance"
+-- (often @Sub Dict@ will do)
+-- Specialized to the case when m ~ Sem r
+absorbWithSem :: forall (p :: (Type -> Type) -> Constraint) -- ^ Monadic Constraint to be absorbed
+                        (x :: (Type -> Type) -> Type -> Type ->Type) -- ^ wrapper to avoid orphan instances
```

Comment:
Missing a space here before the final `Type`

Reply:
I'm also not sure what this kind corresponds to. I'd expect it to be `(Type -> Type) -> Type -> Type`. Where does the extra `Type` come from? Is it the `s` in `MonadState s`?

Reply:
So true.  Fixed.

Reply:
It's the ```s``` you need for Reflection.  The phantom type.